### PR TITLE
Block Hooks API: Add CustomerAccount block to header

### LINKED ISSUE #42757: Block Hooks API: Implement Customer block auto-insertion.
Similar to the work done for the Mini Cart block to auto-insert it into approved themes headers for sites running Woo. We need to investigate whether it's also possible to do this for the Customer block for the same sites.

There may be opportunity during this task to abstract some logic already implemented for the Mini Cart so we can repurpose it for other blocks

### PR DESCRIPTION
**Pull request by tjcafferkey**
### Changes proposed in this Pull Request:

Since https://github.com/WordPress/gutenberg/pull/57754 we're able to insert blocks inside the Navigation block rather than after it. This PR uses the Block Hooks API to auto-insert the Customer Account block inside the navigation for all approved themes and patterns.

Closes https://github.com/woocommerce/woocommerce/issues/42757

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Currently this is only available for WP 6.5 until this PR https://github.com/WordPress/gutenberg/pull/58468 is merged which adds the compatibility layer for earlier WP versions. So some prerequisites:

1. Install [WordPress Beta Tester plugin](https://wordpress.org/plugins/wordpress-beta-tester/) and get latest nightly build.
2. Install Gutenberg `trunk`

**Testing this PR**

1. Test that the Customer Account block is inserted after the Navigation block only in the follow circumstances.
* Only in the Header (not after footer navigations)
* Only if the isn't already a Customer Account block present
2. Test that it displays as "icon only" and doesn't have a label.
3. Test that it appears before the Mini Cart block
4. Test some header patterns.
5. Test that it does not get inserted into TT2 & TT3 themes.
6. Hand pick some other popular block themes since we removed the feature gate (its still excluded to the two themes mentioned above) https://github.com/woocommerce/woocommerce/pull/45581

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add the Customer Account block to the header automatically on approved themes (including TT4).

#### Comment 

</details>

### CODE REVIEW

**Comment by tjcafferkey** [comment]
@nerrad just some thoughts on this issue and a potential solution:

* The anchor blocks parent is a row which has display flex properties. This makes controlling the appearance of how this behaves difficult as you can see in the screenshot. Because of this, I don't think now is the right time to do this (see point below) and we should hold off.
* For WP 6.5 there is an issue to allow [child insertion into the navigation block](https://core.trac.wordpress.org/ticket/59743). Once this is complete (perhaps this is something I could look into?) I think this issue of ours can be solved since inserting the Customer Account block into the navigation would be a more appropriate place to do it.

What do you think?

**Comment by nerrad** [comment]
> For WP 6.5 there is an issue to allow [child insertion into the navigation block](https://core.trac.wordpress.org/ticket/59743). Once this is complete (perhaps this is something I could look into?) I think this issue of ours can be solved since inserting the Customer Account block into the navigation would be a more appropriate place to do it.

It might be a good short-term solution but I anticipate that will lead to some poor starting defaults too:

- if the default menu has a lot of items the customer account item could appear in a weird spot with any wrapping.
- alignment issues could be weird with the customer account block being an icon more tightly aligned with the menu text items (and the mini-cart block way aligned at the right of the container).

What about having some sort of progressive/conditional behaviour, i.e. the insertion point is determined by the attributes of the wrapping container block (or anchor) block? I realize the API might not support that discovery at the moment but it could be a good avenue to explore. It'd be pretty great if auto-insertion dynamically adjusted depending on the theme design and was able to _derive_ that from block layout attributes.

**Comment by tjcafferkey** [comment]
>It might be a good short-term solution but I anticipate that will lead to some poor starting defaults too:

Yeah I completely agree. I think _currently_ if we were to auto-insert the Customer Block in its current state then most users would end up removing (or moving) it especially when they experience it on smaller device viewports.

>What about having some sort of progressive/conditional behaviour, i.e. the insertion point is determined by the attributes of the wrapping container block (or anchor) block?

**Some potential problems I see with this:**
* The hook gets run server-side so if users are playing with their templates/patterns in the editor and meets one of those conditions they won't see the block get auto-inserted until they visit the frontend of the website. This behaviour could lead to frustration and confusion.
* It could become confusing to users (merchants and devs) as to _why_ auto insertion happens with certain themes/patterns/templates and not others.

I _feel_ like we should be auto-inserting blocks regardless (if possible) and they should just work. I keep reminding myself of what Matias said here pdnLyh-4p4-p2#comment-3367. Which is why I thought exploring child insertion into the Navigation block would be a better option as it would _likely_ have a bigger impact since it wouldn't need any conditional behaviour other than checking if the block is already present.

What do you think?

**Comment by nerrad** [comment]
> meets one of those conditions they won't see the block get auto-inserted until they visit the frontend of the website

Auto-insertion isn't supposed to happen on user-edited templates according to the spec for the API, which means this shouldn't be an issue. Or is that something that's changed? 

> It could become confusing to users (merchants and devs) as to why auto insertion happens with certain themes/patterns/templates and not others

Yes, this is a good point, however, what is worse, auto-inserting it impacts the design negatively (making it look like Woo is doing something wrong - when it's more poor design choices with the theme) or not inserting, and the user has to manually add it?

Putting on the user hat, in the context where they might notice this scenario, they are likely trying different themes (or viewing previews) to determine what they'll go with as the design for their store and they _could_ just interpret it as a choice of the theme. In other cases where the user already has an established store and are switching themes, there's likelihood they are already somewhat familiar with being able to customize the design further.

> I feel like we should be auto-inserting blocks regardless (if possible) and they should just work. I keep reminding myself of what Matias said here pdnLyh-4p4-p2#comment-3367. 

...which is why I thought exploring child insertion into the Navigation block would be a better option as it would likely have a bigger impact since it wouldn't need any conditional behaviour other than checking if the block is already present.

Pulling in his comment here because it is relevant:

> We have to assume themes are well built and report back when things break. It’s crucial the we achieve a state of it just works where exceptions are rare.

This comment is one of the reasons why I mentioned I think putting the customer block into the navigation menu has its own set of problems when it comes to layout/design - in other words, I think we'll still find exceptions with that solution as well. I think what Matias is getting at here is that ideally, a theme is designed so that the layout works for injected content so it wraps in reasonable ways - assuming that, any exceptions would be reported to the theme so that it's fixed at the theme level. There might be some best-practices that emerge for theme design so that they better handle auto-insertion from WP plugins via the Block Hooks API.

In the meantime - until we get to the point where the majority or themes "just work", I'm wondering about the user experience and how we could mitigate things so that they aren't forced to do something as a result of the auto-insertion ("ugh I activated Woo and this Customer Account block looks ugly"), and instead can _optionally_ do something ("I activated Woo, I'd like to have an account block next to the mini-cart").

With that in mind, I see a few ways forward here:

- Try the short-term solution you're proposing (anchor block to navigation) across themes (experiment early with the work happening in GB). How does it look across themes, are the concerns I suggested valid or not? Is the breakage more minimal? If it looks promising then it suggest we could roll out with that as a solution.

_If the above isn't promising (or has similar limitations)_

- For the themes with issues, what could be changed about their implementation to make it more friendly to auto-insertion? Is the theme receptive to fixes (that don't significantly change the design, but make it more friendly for auto-insertion)? Maybe we could get the theme fixes pushed upstream for major themes (and any in the DotCom library) and this could help promote best practices that could then be published for all themes as a suggestion to follow.
- Explore the progressive/conditional approach I mentioned. In combination with the above, the places which result in no auto-injection should be few and exceptions would be noticeable (but not user blocking).
- If the previous proves to be too complex, consider always injecting anyways. Assuming we work to improve theme implementations this could end up not being as big of a deal and there'd be enough themes where it works well that it becomes apparent it's a theme issue vs Woo issue.

**Comment by tjcafferkey** [comment]
>Auto-insertion isn't supposed to happen on user-edited templates according to the spec for the API, which means this shouldn't be an issue. Or is that something that's changed?

Sorry, maybe I wasn't clear on my point here. Imagine a scenario where we're conditionally inserting the mini cart based on the attributes of the wrapping block, initially it causes the mini cart not to be inserted. Then the user goes into the Site Editor and starts editing the attributes on the wrapping block which triggers the mini cart to meet the conditions to be inserted. As far as I can tell at this point in the editing experience the mini cart will not appear until they reload the Site Editor but even if they don't reload, without them knowing it is going to be rendering on the frontend after they've saved their changes

>Try the short-term solution you're proposing (anchor block to navigation) across themes (experiment early with the work happening in GB). How does it look across themes, are the concerns I suggested valid or not? Is the breakage more minimal? If it looks promising then it suggest we could roll out with that as a solution.

I'll do some more thorough testing here across different themes and patterns to understand the impact more and if there are issues maybe we can look at one of the alternative solutions you've outlined. I particular am interested in the changes to the themes, because prior to this API it won't have been a consideration when building themes but perhaps now it _should_ be.

**Comment by nerrad** [comment]
> without them knowing it is going to be rendering on the frontend after they've saved their changes

But it won't, because the user has saved the changes? A key point of the Block Hooks API is it won't override user changes. Which means any conditional insertion should only impact _defaults_?

**Comment by tjcafferkey** [comment]
@nerrad ah, you're absolutely right. I'll put this one down to COVID, brain fog can still be used as an excuse right?

**Comment by tjcafferkey** [comment]
@nerrad aside from looking into the issues we've already discussed in this PR I also just want to highlight this exploration done by @ockham here https://github.com/WordPress/wordpress-develop/pull/5837.

This potential solution if accepted could allow us to register a pattern with the Mini Cart & My Account block wrapped in their own Group block which we could then insert as a hooked 'block'.

@ockham I know your PR doesn't exactly solve _your_ layout problem which you're looking into separately but could be beneficial for scenarios where a series of blocks (e.g. Mini Cart and Customer Account blocks) are being inserted into layouts where the parent block (in this case a Row block) has flex properties with various alignment options which at the moment we have no control over. That said though, there may even be a better solution not yet explored for this scenario so just acknowledging that as well.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php]: I think `$VID:$` doesn't work anymore in the monorepo, so we will need to hardcode the WooCommerce version here.

**Review by Aljullu — COMMENTED** [review]
Exciting work on this! :clap: Code-wise, everything looks good to me, but leaving some comments of things I found during testing:

---

When adding any pattern with the Navigation block, the Customer Account block appears twice in the editor. Can you reproduce? The frontend correctly shows only one.

imatge

Sidenote: I also feel it's a bit confusing that the menu which is created when installing WooCommerce contains the _My Account_ page and starting from now there is also the Customer Account block, so in the end there will be two links to the _My Account_ page. That's out of the scope of this project, but it would be nice if in the future WooCommerce came with a default navigation menu that excluded pages like Cart, Checkout and My Account.

---

The Large Footer Dark pattern already has a Customer Account block on the last column. Could we exclude this pattern from autoinserting the My Account block?

imatge

---

Nitpick: should we remove the left and right padding from the Customer Account block when used inside a Navigation block? Otherwise it doesn't look aligned with the other items:

imatge

(Btw, not related to this PR, but wondering if we should add some spacing between the icon and the _My Account_ text, they look too close to me)

**Comment by tjcafferkey** [comment]
> I also feel it's a bit confusing that the menu which is created when installing WooCommerce contains the My Account page and starting from now there is also the Customer Account block, so in the end there will be two links to the My Account page

So you'll probably already know some of this but it is something from WP Core. The default `wp_navigation` (which is used for the Navigation block) contains the inner block `core/page-list` by default which essentially just renders in all the published pages as navigation links. So when Woo is installed it creates these additional pages you mentioned which is necessary, meaning they get rendered with the `core/page-list` block too.

**Comment by tjcafferkey** [comment]
>When adding any pattern with the Navigation block, the Customer Account block appears twice in the editor.

After mine and Alberts call we concluded that `wp-includes/blocks/navigation.php` and Gutenbergs version from the block-library package are both running for him which means the filter is getting hooked _twice_. Strangely its only happening on the editor side and not on the server-side, since code for both exist in the same file.

This requires further investigation as to why it's only happening to him, and if its a wider issue or not.

**Comment by tjcafferkey** [comment]
@Aljullu the issue you found should be fixed in `trunk` Gutenberg now. https://github.com/WordPress/gutenberg/pull/58772

**Review by Aljullu — COMMENTED** [review]
This is looking great, thanks for investigating the tricky issue with the duplicate blocks, Tom! Confirming it's fixed now. :ok_hand: 

A couple of questions:

1. Something I noticed is that if I add the Navigation block in `trunk`, publish and then switch to this branch, the Customer Account block is inserted automatically. Have we considered how this will impact existing stores? Will the Customer Account block be automatically inserted to the navigation of stores on production?
2. That's a minor one, but were you able to reproduce what I mentioned about the _Large Footer Dark_ pattern? Not sure if it's possible to exclude the Customer Account auto-insertion there, but it feels weird that it's added to the navigation columns, considering there is a third column only for the Customer Account block.

**Comment by nerrad** [comment]
Another consideration that comes to mind: If the nav menu already has a link to the customer account page, should the Customer Account block _not_ be inserted?

**Comment by tjcafferkey** [comment]
>Another consideration that comes to mind: If the nav menu already has a link to the customer account page, should the Customer Account block not be inserted?

On fresh installs of WP and WC every page is rendered in the nav since its the fallback menu). IIRC we might be able to do something by checking the context:

* Default sites will just have `<-- core/page-list -->` as an inner block. I _think_ its safe the assume the my account link would be in here since this just renders all published pages (I think?).
* For edited navigations we can inspect each link item since `core/page-list` is replaced with block markup for each page. 

I'll see if its possible to achieve this as part of this PR.

**Comment by tjcafferkey** [comment]
>Something I noticed is that if I add the Navigation block in trunk, publish and then switch to this branch, the Customer Account block is inserted automatically. Have we considered how this will impact existing stores? Will the Customer Account block be automatically inserted to the navigation of stores on production?

Yes thats right, as was the case with inserting the mini block. Not only this but it's just how block hooks should work in general since its the purpose of the API to automatically insert blocks. I guess the trick for us is making sure we do it in a considered way which I think we are.

**Comment by tjcafferkey** [comment]
>That's a minor one, but were you able to reproduce what I mentioned about the Large Footer Dark pattern? Not sure if it's possible to exclude the Customer Account auto-insertion there, but it feels weird that it's added to the navigation columns, considering there is a third column only for the Customer Account block.

We can exclude patterns, allow me to relook at this particular one and I can remove it 😄

**Comment by tjcafferkey** [comment]
>Another consideration that comes to mind: If the nav menu already has a link to the customer account page, should the Customer Account block not be inserted?

@nerrad I've updated the `BlockHooksTrait` to allow custom callbacks for hooked blocks, this shouldn't really be needed much in theory, but it allows us to run some custom block-specific logic which we have to do to solve the above, one of the reasons being that the Navigation blocks implementation of the Block Hooks API is slightly different and tracks ignored blocks via the `wp_postmeta` table.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (tjcafferkey) [inline plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php]: I could change this to become a filter but not sure if I really should as that'll expose it publicly?

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (nerrad) [inline plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php]: A few assumptions I see with this code:

- it _overwrites_ any value in `$hooked_blocks` that may have already been set (instead of appending to the existing array). Was that intentional?
- it assumes that the returned value from the callback is an array. There should probably be type-checking/validation that the expected type is returned from the callback.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (nerrad) [reply plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php]: Yea, can, be added later if the need is demonstrated.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (tjcafferkey) [reply plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php]: >it overwrites any value in $hooked_blocks that may have already been set (instead of appending to the existing array). Was that intentional?

Yeah this was intentional because in some cases we would want to unhook blocks. Like in our specific case in this PR we are checking if the `page-list` block is present, or the `my-account` permalink is, and if so we are unhooking (plus adding the ignore meta). These are really specific conditions to the customer account block which was my motivation for creating this additional callback functionality.

>it assumes that the returned value from the callback is an array. There should probably be type-checking/validation that the expected type is returned from the callback.

I can add that in no problem. I'll also update the testing instructions to cover some of these additional cases before opening it back up for review.

**Comment by tjcafferkey** [comment]
@Aljullu this is ready for review again. I've also excluded the pattern you mentioned.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php:146]: What do you think about adding a comment here explaining why we are excluding these two TT2 patterns?

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php:145]: Is the use case of this filter to allow themes to define patterns so they are excluded from auto-insertion?

One limitation I see is that there is no possibility to decide which specific blocks are auto-inserted or not, right? It's an all-or-nothing approach. I don't think it needs to be changed, but wanted to mention it in case it has been considered.

**Review by Aljullu — COMMENTED** [review]
This is working as described. I like the use of a `callback` to control when to auto-insert the Customer Account block. :clap: 

I did test this with the latest Gutenberg `trunk` and, as mentioned, everything seems to work as expected. I left some comments below with some questions/thoughts. Sorry if I'm asking anything that has been discussed somewhere else, as I have been AFK recently and I might have missed something.

* The Customer Account block is added several times in patterns with several Navigation blocks (see screenshot below), which looks a bit weird. Wondering if it would be possible to auto-insert the Customer Account block only in the first Navigation block rendered on the page or pattern instead of in all of them? :thinking:  I don't know how feasible it is, though.
imatge
* That's a small one, but I think Scenario 2 and Scenario 3 headings are mixed up in the testing steps of the PR description. My understanding is that it should be `Scenario 2 - Edited navigation does not contain My account link` and `Scenario 3 - Edited navigation contains My account link`.

> Yes thats right, as was the case with inserting the mini block. Not only this but it's just how block hooks should work in general since its the purpose of the API to automatically insert blocks. I guess the trick for us is making sure we do it in a considered way which I think we are.

This is more a feature discussion than a technical one, but I wonder if it wouldn't be better to do this only for new installs. I worry that adding one item to the Navigation can break the layout of existing stores in production. While it's an easy problem to solve, as a merchant I wouldn't want to have to check the frontend of my store every time I update WooCommerce to ensure nothing changed.

**Comment by tjcafferkey** [comment]
>This is more a feature discussion than a technical one, but I wonder if it wouldn't be better to do this only for new installs. I worry that adding one item to the Navigation can break the layout of existing stores in production. While it's an easy problem to solve, as a merchant I wouldn't want to have to check the frontend of my store every time I update WooCommerce to ensure nothing changed.

@nerrad @danielwrobert I'm curious to know your thoughts about this? I understand it from both points of view and this question does keep coming up whenever we introduce some enhancements (e.g. classic template to blockified template, block hooks etc).

I have a few thoughts on this in general. If we restrict surfacing enhancements in Woo stores to various conditions then I worry we will end up letting future bugs slip through the net as we create more complex testing environments to assist with coverage for these features.

Also, in the past if a 'Woo specific theme' released updates and stores updated then they would inherit these changes immediately in most cases?

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (tjcafferkey) [reply plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php:145]: Yes that's right. We can make updates to do it on a per-block basis but right now I don't think its required.

**Comment by nerrad** [comment]
I think to a degree that ship has already sailed (at least for the subset of themes we've explicitly supported). I think the _biggest_ concern relates to avoiding duplicate instances of auto-injected blocks and that is already addressed (to a degree) via the block detection that is present? On that note though, I haven't looked at the code - if a block is injected in the navigation, is there detection for the already existing usage of the block in parent blocks?

**Comment by tjcafferkey** [comment]
>if a block is injected in the navigation, is there detection for the already existing usage of the block in parent blocks?

No unfortunately not. The reason for this is because the navigation blocks contents is a `WP_Post` and stored in the database (this is what is passed as `$context` to the API).

We have no idea in what context the navigation is being used so we're unaware of what's 'around' it.

**Comment by nerrad** [comment]
Hmm... so this increases the potential for duplicate blocks on _existing_ sites that have already been customized to have the block right (which in most cases will be outside navigation)?

**Comment by tjcafferkey** [comment]
>Hmm... so this increases the potential for duplicate blocks on existing sites that have already been customized to have the block right (which in most cases will be outside navigation)?

Yeah there is the potential for that. I'm going to spend a little bit of time today on seeing if there is a way using the current implementation to check this, and if so what the trade offs/limitations are of doing so.

**Comment by tjcafferkey** [comment]
An alternative approach would be to look at adding the customer account block after the navigation (before mini cart) but then we have the reverse issue of not knowing if the link is present in the navigation. So I don't think that's feasible either.

**Comment by danielwrobert** [comment]
> An alternative approach would be to look at adding the customer account block after the navigation (before mini cart) but then we have the reverse issue of not knowing if the link is present in the navigation. So I don't think that's feasible either.

If I'm understanding properly (see screenshot below), it _seems_ that this alternative approach would be the better of the two options. For example, while still not ideal, having a Customer Account link the navigation along with a Customer Account block outside of the nav would seem less "broken" than having duplicate Customer Account blocks, IMO.

CleanShot 2024-03-20 at 16 58 18@2x

I'm not necessarily suggesting that we settle on this, mainly noting that it does seem slightly more favorable if we only had those two routes available.

**Comment by tjcafferkey** [comment]
@danielwrobert I'm of the same opinion as you. Currently users cannot add the Customer Account block inside the Navigation block via the editor since its not an allowed block, so there's little risk of duplication there when we're inserting it "after" the navigation block but users may still have the "My account" link there.

That said, today I'm going to time box some investigation to see if theres a way to check the header area for an existence of a block when inserting inside the navigation. I didn't have time yesterday.

File: plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php
Comment (tjcafferkey) [reply plugins/woocommerce/src/Blocks/Utils/BlockHooksTrait.php:146]: https://github.com/woocommerce/woocommerce/pull/43267/commits/83116f9255860d9aed1f8c7b3bafe7b42e1486c4

**Comment by tjcafferkey** [comment]
I did a bit of exploratory work on this with involved:

1. Changing the placement until `after` the Navigation, in the `header` area (this solves the issue Albert raised above where the block was getting hooked into every navigation block in the footer). This would make the `$context` value the header template part.
2. I inspected the context to see if the Navigation had a `ref` value (this is a nav post ID) specified and if so fetched the post. If it didn't I fetched the first `wp_navigation` post type.
3. Checked the inner blocks of this post type to see if there was a "My account" link already there.

I concluded that this wasn't a great solution. I feel like it was too brittle and not reliable given the various fallback mechanisms in place for the navigation block.

Also if Gutenberg changes the attribute value of `ref` to be something other than the post ID for the navigation block (possible, came up during reviews on my work on this block for block hooks) then it would mean we _might_ have to update our code to mirror this.

I'm just going to list out our current options and the pros and cons (along with my thoughts) for us to make a decision:

### Hooking as last_child inside the Navigation block

Pros:
* When the mobile menu collapses the Customer Account block is inside the mobile menu
* We can inspect the inner blocks of the Navigation block to decide if we should/should not hook

Cons:
* Some themes may have the Customer Account block outside the Navigation block, we're not aware of this at the time.
* Everywhere this menu is used the Customer Account block is present. This becomes obvious when you have a 3 column footer each with a Navigation block. Shown here https://github.com/woocommerce/woocommerce/pull/43267#pullrequestreview-1942851596

### Hooking after the Navigation block in the header

Pros: 
* We can specify a header area meaning the Navigation blocks in the footer wont have our Customer Account block present
* On themes that accommodate this well such as TT4 it behaves well on large devices (there are open PRs for TT2/3 to fix this but likely only released in WP 6.6)

Cons:
* On small devices such as mobiles the Customer Account block isnt collapsed behind the menu, by default this is a problem as its too long with the label present.
* We aren't aware if the user has a link to the "my account" page in the navigation because we aren't able to reliably get this information AFAIK.

### My thoughts

By default the current Customer Account block is too long to display "after" the navigation because of the implications of browsing on a mobile device and it not being behind a collapsable menu. **To fix this** we can use the `"hooked_block_{$hooked_block_type}"` filter to set it to hide the label when we hook it into the header.

The only other problem is that of the 4 popular block themes running Woo, TT4 and Tsubaki are the only ones I think that can acceptably accommodate this block being hooked there (see TT3 screenshot below)

Screenshot 2024-03-21 at 13 32 31

I have open PRs for [TT3](https://github.com/WordPress/wordpress-develop/pull/6233) and [TT2](https://github.com/WordPress/wordpress-develop/pull/6234) to fix this for them (see TT3 example below in its current form). Because of this we can exclude these two themes until my PRs have been merged and released.

This only leaves the issue that the users site may already have the My Account link in their Navigation which as far as I can tell we can't do too much about right now.

If we want to use the API to hook this block into merchants sites **my recommendation is to do it after the Navigation block with the changes I've mentioned above** (I will hopefully have pushed them to this PR by the time you've read this).

**Comment by danielwrobert** [comment]
@tjcafferkey great exploration work! I like how you listed out the tradeoffs in each approach as well.

I agree with your proposed solution to hook the block in after the navigation and utilize the `hooked_block_{$hooked_block_type}` filter to improve the mobile experience.

> This only leaves the issue that the users site may already have the My Account link in their Navigation which as far as I can tell we can't do too much about right now.

While this may not be *ideal*, I personally don’t find it to be super concerning. As I noted in my previous comment, it would be more of an issue to have duplicate blocks than it would be to have a link and a block. The latter would look more “broken”, IMO.

Additionally, I believe this issue is also present with the mini cart block where it’s possible that there is a “cart” link in the navigation as well. See the screenshot I included in my last comment for an example.

**Comment by tjcafferkey** [comment]
@Aljullu could you do one final review of the recent changes on this PR please before merging? 🙏🏻

File: plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php]: I like this change. :clap: 

We can now remove the [display style check from `render_label()`](https://github.com/woocommerce/woocommerce/blob/4f2ed8d2627ab53f997b1f1d937f2520e1a98115/plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php#L101-L103), right?

File: plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php]: I might be confused, but now that we are auto-inserting the Customer Account block only in headers, do we still need to exclude this pattern? :thinking:

**Review by Aljullu — APPROVED** [review]
Thanks for updating this PR, @tjcafferkey! I did some testing and this is working as described, the code also looks great. I left just two minor inline comments.

One thing I noticed is that some patterns from the pattern directory look a bit off because the Customer Account and Mini-Cart blocks are not grouped. For example, this is the _Simple header with dark background_ pattern:

imatge

_(above is `trunk`, below is this branch)_

While I don't think this is a blocker for itself, I wonder what would happen in production stores which are currently using this or similar layouts for the header. I know this has already been discussed and this is why I'm approving this PR, but I still see this like a risky change for existing stores in production.

I agree with your concern about existing stores not getting all new features, so I think a middle step would be:

1. Existing stores get the Customer Account block auto-inserted in patterns, but not in the existing header template parts.
2. Newly created stores get the Customer Account block auto-inserted in patterns and the header template part.

I don't know if that's possible or desirable, but I wanted to mention it anyway to know your thoughts.

File: plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php
Comment (tjcafferkey) [reply plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php]: Nice catch. Done in https://github.com/woocommerce/woocommerce/pull/43267/commits/39a8ae3c96051898e056c3135b4229f87440b997

File: plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php
Comment (tjcafferkey) [reply plugins/woocommerce/src/Blocks/BlockTypes/CustomerAccount.php]: Done in https://github.com/woocommerce/woocommerce/pull/43267/commits/e2604fd68d7345091c34f911ac8707979712e6e5

**Comment by tjcafferkey** [comment]
As discussed privately with @Aljullu I'm going to merge this PR as-is (once tests are passing) because we believe the impact of production stores using block themes & the pattern highlighted above as their stores current header is relatively low. Instead we should update the patterns instead of excluding them which is what I'm going to do in this case.

**Pull request merged**
